rfcs/unboxed-types.md

```diff
@@ -0,0 +1,458 @@
+explicitly, e.g.:
+
+    type ('a : value, 'b : immediate) t = Foo of 'a * 'b
+    val fn : ('a : immediate) . 'a ref -> 'a -> unit
+
+In signatures, type variables that are not bound with either of the
+above syntaxes are given layout `value`.
+
+Layout annotations also appear on abstract types:
+
+    type ('a : value, 'b : int32) t : immediate
+
+Again, if this annotation is omitted it's assumed to be `value`.
+
+It may seem a little weird here to allow abstraction over types of
+layout `int32`. After all, there's only one such type (`#int32`).
+However, this ability becomes valuable when combined with abstract
+types: two different modules can expose two different abstract types
+`M.id`, `N.id`, both representing opaque 32-bit identifiers. By
+exposing them as abstract types of layout `#int32`, these modules can
+advertise the fact that the types are represented in four bytes
+(allowing clients of these modules to store them in a memory-efficient
+way), yet still prevent clients from mixing up IDs from different
+modules or doing arithmetic on IDs.
+
+### Unification
+
+As above, type variables are annotated with the layout of the types
+over which they range. During unification, newly-created type
+variables initially range over `any_layout`. Two interesting things
+can then happen: unification of two variables or unification of a
+variable and a non-variable. (Unification of two non-variables works
+as before).
+
+**Two variables** If we have two type variables `'a : L`, `'b : K`,
+then their unifier is a type variable that ranges over only types that
+could be assigned to both `'a` and `'b'`. Its layout is therefore the
+intersection of the layouts `L` and `K`, that is, the layout of types
+that have both layouts. This can be done in general, as layouts form a
+partial semilattice: either there is some layout which is exactly the
+intersection of `L` and `K`, or else the two layouts have no types in
+common (causing a unification error).
+
+**One variable, one non-variable** To unify `'a : L` with a
+non-variable type `t`, we must check first that `t : L`. The
+interesting case here is when `t` is of the form `(p, q) r`, for some
+`r` in the type environment.
+
+There is a subtlety here: the layout of `(p, q) r` can in general depend on
+the layouts of `p` and `q`. For instance:
+
+    type ('a : value) id = 'a
+
+Since `'a : value`, we know that `'a id` is always of layout `value`,
+but if `'a = int` it may also have the more specific layout
+`immediate`.
+
+For each type constructor `(_, _) r` in the environment, we store
+along with it a layout `K` which is an upper bound for the layout of
+any instance of `r`. (For `'a id` above, this is `value`). When
+checking `(p, q) r : L`, we compare `L` and `K`. If `K` is a subtype
+of `L`, then the check passes. However, if the check fails and `r` is
+an alias, we must expand `r` and try again. We only report a
+unification failure once the check fails on a type which is not an
+alias.
```

Comment:
@gasche says
This designs sounds fairly fishy to me:

  - the any_layout kind doesn't really make sense to me
  - the way kinds/layout are mutated during inference is fishy

For example, consider the two definitions

```type ('a : value) id = 'a
type ('a : immediate) immediate = 'a```
And then the following sequences of introduction of new variables and unifications:

```some 'b in (* new: 'b: any_layout *)
some 'c in (* new: 'c: any_layout *)
'b = 'c id;   (* changed: 'b: value, 'c : value *)
'c = immediate int; (* changed: 'c: immediate *)
(* final state: 'b: value, 'c: immediate *)
```
```some 'b in (* new: 'b: any_layout *)
some 'c in (* new: 'c: any_layout *)
'c = immediate int; (* changed: 'c: immediate *)
'b = 'c id;   (* changed: 'b: immediate *)
(* final state: 'b: immediate, 'c: immediate *)
```
The fact that the final state of those two sequences feels wrong to me. I may have misread your algorithm (it's not clear what "the check passes" in term of kind update to the solution), but I think that there are non-principality issues here. The lowering of 'c from any_value to immediate may require the recomputation of any kind decision based on observations of 'c.

Is there a less-convenient but safe proposal that can be used as a first step, to be refined later? Some ideas:

could we consider that only value is ever inferred, and all others must be put explicitly?
what is the design like if we drop immediate, so that there isn't any subtyping of layouts anymore?

@gasche says
If we solve all inference problems in the same way that we are currently doing, I think that we can read back from the solution a minimal classification of type variables into layouts -- or fail with an error if no such assignment exists. Is that right?

The other natural approach, I think, would be to be fully explicit about layout variables and their resolution during inference -- much like the explicit handling of universe levels in dependently-typed programming language. At the level of types, we use unification/equality as the base mechanism, but you are proposing to do inference with subtyping at the level of layouts. I would like to see a more explicit description of how that works (or another approach that avoids this sophistication) than the fairly imprecise description in the current document.

@lpw25 says
> the any_layout kind doesn't really make sense to me

I believe an equivalent proposal is as follows:

Have a notion of "sort" that classifies the layouts by compatibility -- one sort for value and immediate, one sort for int64, etc.
Add family of layouts Top of sort which are the top of the subtyping order within each sort. For most of the layouts this will be the only layout in the lattice (e.g. Top of the int64 sort would be equal to the int64 layout).
Allow sort unification variables which cannot be generalized
Then you can replace all uses of any_layout with Top 's (where 's is a sort unification variable)
When generalizing a type only allow it to be generalized if it's layout does not contain any sort variables
So maybe that helps to clarify things. Stephen and I somewhat disagree over this part, I think that using sort variables and then removing the last restriction -- which makes layout inference be principal -- is a cleaner system. However, since my version is backwards compatible with his, I think his version is a reasonable one for an initial prototype.

@stedolan says
@gasche I think you did misread the algorithm! I'll see if I can find some time to write it out more clearly and precisely. The algorithm is really a small change to classical unification!

Type variables have layouts, describing what they range over. Once a type variable is instantiated (= substituted away), we no longer have either a variable or its layout, just the result of substituting. In your example:
```
some 'b in (* new: 'b: any_layout *)
some 'c in (* new: 'c: any_layout *)
'c id (* to form the type 'c id, 'c is restricted to value *)
'b = 'c id (* substitute 'c id for 'b *)
(* state is: 'c : value; 'b = 'c id *)
'c = int (* substitute int for 'c *)
(* final state is 'b = 'c id; 'c = int *)
```
or the other way around:

```some 'b in (* new: 'b: any_layout *)
some 'c in (* new: 'c: any_layout *)
'c = int (* substitute int for 'c *)
(* state is: 'b: any_layout; 'c = int *)
'b = 'c id
(* final state is 'c = int; 'b = 'c id *)
```
The design is indeed somewhat simpler if we drop immediate: we can solve layouts by standard unification, with a restriction like the any_layout restriction here on when they can be generalised.

I think we need to consider immediate now, though: there's clearly a demand for such features ([@@immediate] is part of the language today), and grafting subtyping onto an inference system afterwards does not tend to be straightforward.

 @Drup says
I much prefer @lpw25 's version, which I suspect is a strip down version of what I did for linear types. I feel like having sort variables will serve us far better in the larger context.